--!nocheck

local ItemBook = {}

ItemBook.Items = {
    {
        ["ItemId"] = "HealthyApple",
        ["Name"] = "Healthy Apple",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "A very healthy apple from a healthy tree",
        ["Stat"] = "Recovers <green>10% Health</font>",
        ["Icon"] = 1637563032,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 5
    },
    {
        ["ItemId"] = "GoldenApple",
        ["Name"] = "Golden Apple",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A gift for those who are blessed by gods",
        ["Stat"] = "Recovers <green>35% Health</font>, and gives you <blue>1.5X Attack</font> buff for <blue>5 seconds</font>",
        ["Icon"] = 2970814988,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "HealthPotion",
        ["Name"] = "Health Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Recovers <green>30% Health</font>",
        ["Icon"] = 2970816690,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 15
    },
    {
        ["ItemId"] = "PowerPotion",
        ["Name"] = "Power Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Gives you <blue>2X Attack</font> buff for <blue>8 seconds</font> ",
        ["Icon"] = 2970818097,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "MagicPotion",
        ["Name"] = "Magic Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Gives you <purple>2X Magic</font> buff for <blue>8 seconds</font>",
        ["Icon"] = 2970817110,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "ManaPotion",
        ["Name"] = "Mana Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Recovers <blue>30% Mana</font>",
        ["Icon"] = 2970816149,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "PurifyPotion",
        ["Name"] = "Purify Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Cures off bad status effects",
        ["Icon"] = 2970818438,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "FireballScroll",
        ["Name"] = "Ignite Scroll",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A scroll that found in the ancient",
        ["Stat"] = "Deals <purple>half magic damage</font> to the enemy and inflicts <red>Burn</font>, costs <blue>50 Mana</font>",
        ["Icon"] = 12027128349,
        ["IsSkill"] = true,
        ["Mana"] = 50,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "Agility",
        ["Name"] = "Agility",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "A skill that can be learned easily",
        ["Stat"] = "Speed you up for <blue>6s</font>",
        ["Icon"] = 4824421210,
        ["IsSkill"] = true,
        ["Mana"] = 0,
        ["Price"] = 500
    },
    {
        ["ItemId"] = "ExtraJump",
        ["Name"] = "Extra Jump",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "Great for exploring",
        ["Stat"] = "Boost you up!",
        ["Icon"] = 4821548100,
        ["IsSkill"] = true,
        ["Mana"] = 0,
        ["Price"] = 500
    },
    {
        ["ItemId"] = "BloxyCola",
        ["Name"] = "Bloxy Cola",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Refreshing sweet soda, which would give you a sugar rush",
        ["Stat"] = "Gives you <blue>1.5X Speed</font> buff for <blue>8 seconds</font>, and recovers <green>15% Health</font>",
        ["Icon"] = 4865339572,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "CookedSteak",
        ["Name"] = "Cooked Steak",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Well cooked and tasty",
        ["Stat"] = "Gives you <blue>1.2X Attack</font> buff for <blue>6 seconds</font>, and recovers <green>10% Health</font>",
        ["Icon"] = 17481500518,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 30
    },
    {
        ["ItemId"] = "ArcaneMark",
        ["Name"] = "Arcane Mark",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A scroll that used by great magicians",
        ["Stat"] = "Fires a ball towards enemy, it will mark an area under it, which deals <purple>magic damage</font> five times if hits, costs <blue>100 Mana</font>",
        ["Icon"] = 14429505922,
        ["IsSkill"] = true,
        ["Mana"] = 100,
        ["Price"] = 1000
    },
    {
        ["ItemId"] = "FoxFlameScroll",
        ["Name"] = "Fox Scroll",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "Kitsunes always carry these scrolls, just to find a new master",
        ["Stat"] = "Deals <purple>decent magic damage</font> to the enemy and inflicts <red>Burn</font>, costs <blue>50 Mana</font>",
        ["Icon"] = 4820482325,
        ["IsSkill"] = true,
        ["Mana"] = 50,
        ["Price"] = 1000
    },
    {
        ["ItemId"] = "GhostlyLantern",
        ["Name"] = "Ghostly Lantern",
        ["Type"] = "Active",
        ["Tier"] = 3,
        ["Desc"] = "Ghosts, hear my commands!",
        ["Stat"] = "Deals <purple>magic damage</font> to the enemy three times and recovers <green>20% Health</font>, costs <blue>200 Mana</font>",
        ["Icon"] = 4820486105,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 1500
    },
    {
        ["ItemId"] = "HolySmite",
        ["Name"] = "Holy Smite",
        ["Type"] = "Active",
        ["Tier"] = 3,
        ["Desc"] = "A spell that calls the smite from Heaven",
        ["Stat"] = "Strikes down a beam to deal <purple>magic damage</font> to the enemy and inflicts <red>BURN</font> for <blue>6 seconds</font>, costs <blue>200 Mana</font>",
        ["Icon"] = 14429523818,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 2000
    },
    {
        ["ItemId"] = "WarriorsPride",
        ["Name"] = "Warrior's Pride",
        ["Type"] = "Active",
        ["Tier"] = 3,
        ["Desc"] = "A great pride for Warriors",
        ["Stat"] = "Slashs out a blade of pure raw physical energy, which deals <red>physical damage</font>, costs <blue>200 Mana</font>",
        ["Icon"] = 17355782628,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 15000
    },
    {
        ["ItemId"] = "TravellerBoots",
        ["Name"] = "Traveller's Boots",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "A pair of comfortable boots for Adventurers",
        ["Stat"] = "Makes you slightly faster in overworld",
        ["Icon"] = 2965728606,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "BurningTorch",
        ["Name"] = "Burning Torch",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "A magical torch which even burns in rain",
        ["Stat"] = "Emits the firelight wherever you are, and make you slightly faster in overworld",
        ["Icon"] = 2965771738,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "CampfireWood",
        ["Name"] = "Campfire Wood",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "Makes the campfire bigger!",
        ["Stat"] = "Heals <green>30% more health</font> when resting at a campfire",
        ["Icon"] = 2970559088,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "HuntingDagger",
        ["Name"] = "Hunting Dagger",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "A sharp dagger for hunting foxes",
        ["Stat"] = "<red>+5 Min Damage</font>, simple",
        ["Icon"] = 1723200131,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "MiningPickaxe",
        ["Name"] = "Mining Pickaxe",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "A pickaxe for mining ores",
        ["Stat"] = "Allows you to mine <blue>Iron Ore</font> in the cave",
        ["Icon"] = 12118509077,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "CopperSword",
        ["Name"] = "Copper Sword",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "You feel like it contains a long history",
        ["Stat"] = "<red>+15 Max Damage</font>, simple",
        ["Icon"] = 12192841092,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "AncientHammer",
        ["Name"] = "Ancient Hammer",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "You feel those many battles it's been through",
        ["Stat"] = "<red>+20 Min Damage</font>, <red>+40 Max Damage</font>, and <blue>+20 Shield</font>",
        ["Icon"] = 12119868315,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ManaCrystal",
        ["Name"] = "Mana Crystal",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "A raw mana crystal, which contains some mana in it",
        ["Stat"] = "Restores <blue>20% Mana</font> in every 15 seconds, and <blue>+100 Mana</font>",
        ["Icon"] = 12119869276,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "RabbitFoot",
        ["Name"] = "Rabbit Foot",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "Originally it's a cosmetic?",
        ["Stat"] = "<blue>+20 Jump</font>, a good passive for exploring the world",
        ["Icon"] = 4820619821,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ArcaneSpellbook",
        ["Name"] = "Arcane Spellbook",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "A magical spellbook used by the Life Faction members",
        ["Stat"] = "<purple>+120 Magic</font>, <blue>+100 Mana</font> and be a member of <orange>Life Faction</font>",
        ["Icon"] = 17286830798,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "MagicianHat",
        ["Name"] = "Magician's Hat",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "A hat wore by the Life Faction members",
        ["Stat"] = "<purple>+120 Magic</font>, <green>+300 Health</font>, and be a member of <orange>Matter Faction</font>",
        ["Icon"] = 14429547149,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "EnchantedBlade",
        ["Name"] = "Enchanted Blade",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "A sword blessed by magic!",
        ["Stat"] = "<red>+25 Min Damage</font>, <red>+100 Max Damage</font>, and <purple>+50 Magic</font>",
        ["Icon"] = 4820519895,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "StalkerCape",
        ["Name"] = "Stalker Cape",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "A cape made out of stalker leaves",
        ["Stat"] = "<blue>+15% Dodge Chance</font>, <blue>+5 Speed</font>, and <green>+100 Health</font>",
        ["Icon"] = 12284033809,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "RegenerationGem",
        ["Name"] = "Regeneration Gem",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "A gem that can warmth your heart",
        ["Stat"] = "Heals <green>10% Health</font> in every 15 seconds, and <green>+150 Health</font>",
        ["Icon"] = 17481477951,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "BlackwolfNecklace",
        ["Name"] = "Black Wolf Necklace",
        ["Type"] = "Passive",
        ["Tier"] = 4,
        ["Desc"] = "It's your glory!",
        ["Stat"] = "<red>+75 Min Damage</font>, <red>+150 Max Damage</font>, and <blue>+10 Speed</font>",
        ["Icon"] = 17481485403,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "FlameSword",
        ["Name"] = "Flame Sword",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "Sharp, and hot, perfect for battle",
        ["Stat"] = "<red>+10% Burn Chance</font>, <purple>+25 Magic</font>, and <red>+100 Max Damage</font>",
        ["Icon"] = 17481496050,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 1000
    },
    {
        ["ItemId"] = "KitsuneGloves",
        ["Name"] = "Kitsune Gloves",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "It warmth your hands",
        ["Stat"] = "<purple>+50 Magic</font>, and <red>+50 Max Damage</font>",
        ["Icon"] = 12285229661,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "AquaLapis",
        ["Name"] = "Aqua Lapis",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "An incredibly rare stone with magical properties based off water",
        ["Stat"] = "<purple>+30 Magic</font>, <blue>+150 Mana</font>",
        ["Icon"] = 17515925140,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "EverfrostSeed",
        ["Name"] = "Everfrost Seed",
        ["Type"] = "Passive",
        ["Tier"] = 5,
        ["Desc"] = "An identification of being worthy to enter their stronghold",
        ["Stat"] = "Allows you to enter <blue>Northern Stronghold</font>",
        ["Icon"] = 14429517439,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "HotPotato",
        ["Name"] = "Hot Potato",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "Not actually a potato, just a piece of charcoal that has absorbed a lot of heat",
        ["Stat"] = "<red>+10% Burn Chance</font>, <blue>+15 Speed</font>",
        ["Icon"] = 4820525274,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "NorthernerAxe",
        ["Name"] = "Northerner Axe",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "Used by soldiers from the north",
        ["Stat"] = "<red>+70 Max Damage</font>, and allows you to cut trees in certain areas",
        ["Icon"] = 14429550726,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "TundraGear",
        ["Name"] = "Tundra Gear",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "Used by hikers and adventurers",
        ["Stat"] = "Allows you to immense the cold in some areas of <blue>Tundra</font>",
        ["Icon"] = 2985521957,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 10000
    },
    {
        ["ItemId"] = "FoxSpirit",
        ["Name"] = "Fox Spirit",
        ["Type"] = "Passive",
        ["Tier"] = 4,
        ["Desc"] = "This is, the spirit of Kitsune, and their future",
        ["Stat"] = "<purple>+50 Magic</font>, and <blue>+15 Speed</font> with the aid of Kitsune Spirit.",
        ["Icon"] = 4820522841,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 3000
    },
    {
        ["ItemId"] = "NorthernHalberd",
        ["Name"] = "Northern Halberd",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "A great halberd forged within the Northern Stronghold",
        ["Stat"] = "<red>+40 Min Damage</font>, <red>+80 Max Damage</font>",
        ["Icon"] = 4827596848,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "AbsorptionCrystal",
        ["Name"] = "Absorption Crystal",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "A magical crystal taken from deep below from the caverns",
        ["Stat"] = "<orange>+50 Shield</font>, <blue>+50 Mana</font>, restores <blue>10% more Mana</font> for every attack",
        ["Icon"] = 4820513616,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 2000
    },
    {
        ["ItemId"] = "FireStaff",
        ["Name"] = "Fire Staff",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "This little girl cannot explain this staff",
        ["Stat"] = "Summons a fire staff orb in every 7 seconds in combat, which inflicts <red>Burn</font> for every orb attack",
        ["Icon"] = 4820522220,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 5000
    },
    {
        ["ItemId"] = "MountainKingCrown",
        ["Name"] = "Mountain King's Crown",
        ["Type"] = "Passive",
        ["Tier"] = 5,
        ["Desc"] = "A crown that belongs to an ancient king, well he is dead, now soooo grave robbing?",
        ["Stat"] = "<green>+300 Health</font>, <orange>+60 Shield</font>, and <red>+100 Max Damage</font>",
        ["Icon"] = 4820946334,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 10000
    },
    {
        ["ItemId"] = "SlimeCore",
        ["Name"] = "Slime Core",
        ["Type"] = "Material",
        ["Tier"] = 1,
        ["Desc"] = "The core of slimes, a good material for alchemy",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 12118510527,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "SlimeGel",
        ["Name"] = "Slime Gel",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "The gel of slimes, sticky",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 2980409698,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "SpikyFur",
        ["Name"] = "Spiky Fur",
        ["Type"] = "Material",
        ["Tier"] = 1,
        ["Desc"] = "The alternative version of Kitsune Fur",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 12118511428,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "SpikyQuills",
        ["Name"] = "Spiky Quills",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Nice material for crafting",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 2980409932,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "StalkerLog",
        ["Name"] = "Stalker Log",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "It has stalker energy in it",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 12159550747,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "StalkerLeaf",
        ["Name"] = "Stalker Leaf",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Stalker makes them more lush",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 17481492369,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "KitsuneFur",
        ["Name"] = "Kitsune Fur",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "A soft fur that has magical coating within it",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 12191831926,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "KitsuneFlame",
        ["Name"] = "Kitsune Flame",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "The inner spirit of Kitsune",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 14429538456,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "IronOre",
        ["Name"] = "Iron Ore",
        ["Type"] = "Material",
        ["Tier"] = 1,
        ["Desc"] = "The classic metal that everyone in the world has used",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 14429532190,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "BlackFur",
        ["Name"] = "Black Fur",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Manifesting negative energy",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 12285144842,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "HeheLetter",
        ["Name"] = "A Letter From Hehe",
        ["Type"] = "Quest",
        ["Tier"] = 2,
        ["Desc"] = "Without any scent on it, after all, it's just the most ordinary letter",
        ["Stat"] = "A recommendation letter from hehe, which helps your journey to <blue>Northern Stronghold</font>.",
        ["Icon"] = 16654291188,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ArmorBreaker",
        ["Name"] = "Armor Breaker",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A spell that is famously used for breaking hard carapaces or metals",
        ["Stat"] = "Deals <red>max physical damage</font> to the enemy regardless of defense, costs <blue>50 Mana</font>",
        ["Icon"] = 4820467850,
        ["IsSkill"] = true,
        ["Mana"] = 50,
        ["Price"] = 3000
    },
    {
        ["ItemId"] = "Bomb",
        ["Name"] = "Bomb",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "Powerful weapons created to break through the hardest defenses",
        ["Stat"] = "Deals <red>3X min physical damage</font> to the enemy, costs <blue>50 Mana</font>",
        ["Icon"] = 4824414829,
        ["IsSkill"] = true,
        ["Mana"] = 50,
        ["Price"] = 5000
    },
    {
        ["ItemId"] = "DivineProtection",
        ["Name"] = "Divine Protection",
        ["Type"] = "Active",
        ["Tier"] = 4,
        ["Desc"] = "A skill that calls out protection from the gods",
        ["Stat"] = "<blue>+90% Defense</font> for <blue>6 seconds</font>, costs <blue>100 Mana</font>",
        ["Icon"] = 4820480587,
        ["IsSkill"] = true,
        ["Mana"] = 100,
        ["Price"] = 10000
    },
    {
        ["ItemId"] = "EndurancePotion",
        ["Name"] = "Endurance Potion",
        ["Type"] = "Active",
        ["Tier"] = 1,
        ["Desc"] = "Created by great alchemists in Magic Faction",
        ["Stat"] = "Gains <blue>10% Defense</font> for <blue>6 seconds</font>",
        ["Icon"] = 2970818737,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 10
    },
    {
        ["ItemId"] = "GoldenGel",
        ["Name"] = "Golden Gel",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Tastes like honey",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 4824417771,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "HealingSpell",
        ["Name"] = "Healing Spell",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A spell that can heal you",
        ["Stat"] = "Recovers <green>50% Health</font> for your teammates, costs <blue>200 Mana</font>",
        ["Icon"] = 12027131895,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 3000
    },
    {
        ["ItemId"] = "IceCandy",
        ["Name"] = "Ice Candy",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "Sweet treats from the north that came from slimes",
        ["Stat"] = "Recovers <green>66% Health</font>",
        ["Icon"] = 4806921700,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 15
    },
    {
        ["ItemId"] = "CheeseBurger",
        ["Name"] = "Cheese Burger",
        ["Type"] = "Active",
        ["Tier"] = 5,
        ["Desc"] = "Kitty, you can have this burger. Smells yummy, I wonder whats the recipe?",
        ["Stat"] = "Recovers <green>100% Health</font>",
        ["Icon"] = 4827591686,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 1000
    },
    {
        ["ItemId"] = "Tenebris",
        ["Name"] = "Tenebris",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A spell that casts a ball of darkness",
        ["Stat"] = "Deals tremendous <purple>magic damage</font> to the enemy, costs <blue>200 Mana</font>",
        ["Icon"] = 4824383801,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 2000
    },
    {
        ["ItemId"] = "Lux",
        ["Name"] = "Lux",
        ["Type"] = "Active",
        ["Tier"] = 2,
        ["Desc"] = "A spell that casts a ball of light",
        ["Stat"] = "Deals heavy <purple>magic damage</font> to the enemy, costs <blue>200 Mana</font>",
        ["Icon"] = 4820492332,
        ["IsSkill"] = true,
        ["Mana"] = 200,
        ["Price"] = 2000
    },
    {
        ["ItemId"] = "Devilsknife",
        ["Name"] = "Devisknife",
        ["Type"] = "Passive",
        ["Tier"] = 2,
        ["Desc"] = "Deltarune?",
        ["Stat"] = "<purple>+30% Poison Chance</font>",
        ["Icon"] = 2970815325,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 5000
    },
    {
        ["ItemId"] = "SharpenerRock",
        ["Name"] = "Sharpener Rock",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "A flat rock used to sharpen blades",
        ["Stat"] = "<yellow>+10% Critical Chance</font>",
        ["Icon"] = 2965825317,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "BatWing",
        ["Name"] = "Bat Wing",
        ["Type"] = "Material",
        ["Tier"] = 1,
        ["Desc"] = "Dropped by Bats, bats make me crazy",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 2970814260,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "CrimsonSlimeCore",
        ["Name"] = "Crimson Slime Core",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Dropped by Crimson Slime, it's definitely more annoying than Purple Slime",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 4785691629,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "Ectoplasm",
        ["Name"] = "Ectoplasm",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Ghosts are made out of these weird odd substances",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 4824434937,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "GoldOre",
        ["Name"] = "Gold Ore",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Expensive rare shiny material",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4797656231,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "IceSlimeCore",
        ["Name"] = "Ice Slime Core",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Dropped by Ice Slime",
        ["Stat"] = "Used for <purple>Alchemy</font>",
        ["Icon"] = 4785692076,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "LurkerClaws",
        ["Name"] = "Lurker Digging Claws",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "I wonder ifбн",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4801461630,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "LurkerShell",
        ["Name"] = "Lurker Shell",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Must be incredibly hard to fight these things",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4821138900,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "MythrilOre",
        ["Name"] = "Mythril Ore",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Expensive rare ore used to create strong metals",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4797656828,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ElderWolfFur",
        ["Name"] = "Elder Wolf Fur",
        ["Type"] = "Material",
        ["Tier"] = 2,
        ["Desc"] = "Dropped by the Elder Wolf, an endangered species.",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4785692666,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ElderWolfQuills",
        ["Name"] = "Elder Wolf Quills",
        ["Type"] = "Material",
        ["Tier"] = 3,
        ["Desc"] = "Dropped by the Elder Wolf, an endangered species.",
        ["Stat"] = "Used for <orange>Crafting</font>",
        ["Icon"] = 4785693140,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "AdventurersKit",
        ["Name"] = "Adventurer's Kit",
        ["Type"] = "Passive",
        ["Tier"] = 3,
        ["Desc"] = "a trusty kit useful for adventuring, and more usefully, killing monsters",
        ["Stat"] = "Spawns <blue>2 orbs</font> instantly whenever you enter combat, heals <green>5%</font> more in campfire, and restores <blue>5%</font> more Mana from orbs",
        ["Icon"] = 12370527641,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "Amulet",
        ["Name"] = "Amulet",
        ["Type"] = "Passive",
        ["Tier"] = 5,
        ["Desc"] = "An amulet made from gold, maybe you can try to add something to it",
        ["Stat"] = "It's just good, no extra stats will be given",
        ["Icon"] = 4834726811,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ClimbingGear",
        ["Name"] = "Climbing Gear",
        ["Type"] = "Passive",
        ["Tier"] = 1,
        ["Desc"] = "Used by hikers and adventurers",
        ["Stat"] = "Allows you to climb certain walls",
        ["Icon"] = 4820516087,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "CrimsonSlimeFang",
        ["Name"] = "Crimson Slime Fang",
        ["Type"] = "Passive",
        ["Tier"] = 4,
        ["Desc"] = "Fangs originating from a dangerous type of slime",
        ["Stat"] = "<red>+50 Min Damage</font>, <red>+100 Max Damage</font>, <blue>+5 Jump</font>, and <yellow>+10% Critical Chance</font>",
        ["Icon"] = 4820517734,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 0
    },
    {
        ["ItemId"] = "ValkyrieArmor",
        ["Name"] = "Valkyrie",
        ["Type"] = "Cosmetic",
        ["Tier"] = 4,
        ["Desc"] = "Forged from adamantite and orichalcum along with enchantments of light",
        ["Stat"] = "Used by the valkyries from Valhalla",
        ["Icon"] = 14429563981,
        ["IsSkill"] = false,
        ["Mana"] = 0,
        ["Price"] = 50000
    }
}

ItemBook.ItemImagesEx = {
    Null = "rbxassetid://1637592633",
    Unknown = "rbxassetid://2985521957",

    Gold = "rbxassetid://18421784947",
    RP = "rbxassetid://16002068286",
    Backpack = "rbxassetid://2970814599"
}

--[[
    Find a item through the list
]]
function ItemBook:IsItemExist(item: string) : boolean
    local check = false

    for _, itemTable: table in ItemBook.Items do
        if itemTable.ItemId == item then check = true break end
    end

    return check
end

--[[
    Find a item, and then return it
]]
function ItemBook:FindItem(item: string)
    local itemInfo = {}

    for _, itemTable: table in ItemBook.Items do
        if itemTable.ItemId == item then itemInfo = itemTable break end
    end

    return itemInfo :: {
        ItemId: string,
        Name: string,
        Type: string,
        Tier: string,
        Desc: string,
        Stat: string,
        Icon: number,
        IsSkill: boolean,
        Mana: number,
        Price: number
    }
end

return ItemBook